### PR TITLE
feat(sca): add extract nodes igraph

### PR DESCRIPTION
**Pull request by lirshindalman**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

*Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.*

Fixes # (issue)

### Description
*Include a description of what makes it a violation and any relevant external links.*

### Fix
*How does someone fix the issue in code and/or in runtime?*

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/common/images/graph/image_referencer_provider.py [hunk 1]
```diff
@@ -46,4 +47,24 @@ def extract_nodes_networkx(self) -> networkx.Graph:
...
-        pass
+        resource_nodes = [
+            node
+            for node, resource_type in zip(self.graph_connector.vs['name'], self.graph_connector.vs['resource_type'])
+            if resource_type and resource_type in self.supported_resource_types
+        ]
+        return self.graph_connector.subgraph(resource_nodes)
+
+    def extract_resource(self, supported_resources_graph: networkx.Graph | igraph.Graph) -> Generator[dict[str, Any], None, None]:
+        if self.graph_framework == 'NETWORKX':
```
Comment (tronxd) [inline checkov/common/images/graph/image_referencer_provider.py:58]: Extract the 2 implementations into separate methods

File: checkov/common/images/graph/image_referencer_provider.py [hunk 2]
```diff
@@ -46,4 +47,24 @@ def extract_nodes_networkx(self) -> networkx.Graph:
         return self.graph_connector.subgraph(resource_nodes)
 
     def extract_nodes_igraph(self) -> igraph.Graph:
-        pass
+        resource_nodes = [
+            node
+            for node, resource_type in zip(self.graph_connector.vs['name'], self.graph_connector.vs['resource_type'])
```
Comment (tronxd) [inline checkov/common/images/graph/image_referencer_provider.py:52]: Use `CustomAttributes.RESOURCE_TYPE`

File: checkov/common/images/graph/image_referencer_provider.py [hunk 3]
```diff
@@ -46,4 +47,24 @@ def extract_nodes_networkx(self) -> networkx.Graph:
...
+    def extract_resource(self, supported_resources_graph: networkx.Graph | igraph.Graph) -> Generator[dict[str, Any], None, None]:
+        if self.graph_framework == 'NETWORKX':
+            for _, resource in supported_resources_graph.nodes(data=True):
+                yield resource
+        elif self.graph_framework == 'IGRAPH':
+            for v in supported_resources_graph.vs:
+                resource = {
+                    'name': v['name'],
+                    'block_type_': v['block_type_'],
+                    'resource_type': v['resource_type']
```
Comment (tronxd) [inline checkov/common/images/graph/image_referencer_provider.py:66]: Use `CustomAttributes.RESOURCE_TYPE` and `Use CustomAttributes.BLOCK_TYPE`

File: checkov/common/images/graph/image_referencer_provider.py [hunk 4]
```diff
@@ -30,11 +29,13 @@ def __init__(self, graph_connector: Union[igraph.Graph, networkx.DiGraph],
...
 
-    def extract_nodes(self) -> networkx.Graph | igraph.Graph:
+    def extract_nodes(self) -> networkx.Graph | igraph.Graph | None:
         if self.graph_framework == 'NETWORKX':
             return self.extract_nodes_networkx()
-        else:
+        elif self.graph_framework == 'IGRAPH':
             return self.extract_nodes_igraph()
+        else:
+            return None
```
Comment (tronxd) [inline checkov/common/images/graph/image_referencer_provider.py:38]: Do we handle the case where the the returned subgraph is `None`?

File: tests/cloudformation/image_referencer/test_manager.py [hunk 5]
```diff
@@ -24,8 +32,19 @@ def test_extract_images_from_resources():
...
+        graph = DiGraph()
+        graph.add_node(1, **aws_resource)
+    else:
+        graph = igraph.Graph()
+        attr = aws_resource
+        graph.add_vertex(
+            name='1',
+            block_type_='resource',
+            resource_type=attr[CustomAttributes.RESOURCE_TYPE] if CustomAttributes.RESOURCE_TYPE in attr else None,
+            attr=attr,
```
Comment (tronxd) [inline tests/cloudformation/image_referencer/test_manager.py:45]:
```suggestion
            attr=aws_resource,
```

File: tests/cloudformation/image_referencer/test_manager.py [hunk 6]
```diff
@@ -24,8 +32,19 @@ def test_extract_images_from_resources():
...
         "resource_type": "AWS::ECS::TaskDefinition",
     }
-    graph = DiGraph()
-    graph.add_node(1, **aws_resource)
+    if graph_framework == 'NETWORKX':
+        graph = DiGraph()
+        graph.add_node(1, **aws_resource)
+    else:
+        graph = igraph.Graph()
+        attr = aws_resource
```
Comment (tronxd) [inline tests/cloudformation/image_referencer/test_manager.py:40]:
```suggestion
```

File: checkov/cloudformation/image_referencer/base_provider.py [hunk 7]
```diff
@@ -16,7 +16,7 @@ def extract_images_from_resources(self) -> list[Image]:
 
         supported_resources_graph = self.extract_nodes()
 
-        for _, resource in supported_resources_graph.nodes(data=True):
+        for resource in self.extract_resource(supported_resources_graph):
```
Comment (tronxd) [inline checkov/cloudformation/image_referencer/base_provider.py:19]: Please fill all other providers other than `cloudformation`

File: checkov/common/images/graph/image_referencer_provider.py (hunk 4 above)
Comment (lirshindalman) [reply checkov/common/images/graph/image_referencer_provider.py:38]: the default value of the graph framework is 'NETWORKX', so no. change the function and the 'else' will be the default NETWORKX

**Review by tronxd — APPROVED** [review]

**Review by Eliran-Turgeman — APPROVED** [review]

**Pull request merged**